jax/_src/numpy/lax_numpy.py

```diff
@@ -4128,30 +4128,20 @@ def eye(N: DimSize, M: DimSize | None = None,
            [0., 0., 1., 0., 0.],
            [0., 0., 0., 1., 0.]], dtype=float32)
   """
-  # TODO(vfdev-5): optimize putting the array directly on the device specified
-  # instead of putting it on default device and then on the specific device
-  output = _eye(N, M=M, k=k, dtype=dtype)
-  if device is not None:
-    return jax.device_put(output, device=device)
-  return output
-
-
-def _eye(N: DimSize, M: DimSize | None = None,
-        k: int | ArrayLike = 0,
-        dtype: DTypeLike | None = None) -> Array:
   dtypes.check_user_dtype_supported(dtype, "eye")
   if isinstance(k, int):
     k = lax_internal._clip_int_to_valid_range(k, np.int32)
   util.check_arraylike("eye", k)
-  offset = asarray(k)
+  offset = asarray(k, device=device)
```

Comment:
Does `offset` need to be on the `device`?

Reply:
as we do `i + offset` in the last statement, I was thinking this would prevent moving data between devices

Reply:
But `device` may be a `sharding` (which will match the output shape), and `offset` is a scalar, so it won't work in the general case.